src/clj/puppetlabs/services/jruby_pool_manager/impl/jruby_agents.clj

```diff
@@ -86,87 +105,101 @@
    pool :- jruby-schemas/pool-queue-type]
   (= expected-pool-size (count (.getRegisteredElements pool))))
 
-(schema/defn ^:always-validate
-  swap-and-drain-pool!
-  "Replace the current pool with a new pool and drain the old pool,
-  optionally refilling the new pool with fresh jrubies."
+(schema/defn collect-all-jrubies :- [JRubyInstance]
+  "Locks the pool and borrows all the instances"
+  [pool-state :- jruby-schemas/PoolState]
+  (let [pool-size (:size pool-state)
+        pool (:pool pool-state)
+        borrow-fn (fn [_] (jruby-internal/borrow-from-pool!*
+                           jruby-internal/borrow-without-timeout-fn
+                           pool))]
+    (.lock pool)
+    (try
+      (mapv borrow-fn (range pool-size))
+      (catch Exception e
+        (.clear pool)
+        (.insertPill pool (PoisonPill. e))
+        (throw (IllegalStateException.
```

Comment:
Yeah, I think it might make sense to store it in both places - maybe as a peer to `:modify-instance-agent` in the map associated with the `:internal` key in the pool context and as state in the agent itself.  Seems convenient to keep it in the agent state so that the receiving function in the agent can use it.